pentaho-proc-governor/impl/src/main/java/org/hitachivantara/process/ProcessGovernorImpl.java

```diff
@@ -0,0 +1,134 @@
+/*!
+ * Copyright 2010 - 2018 Hitachi Vantara.  All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package org.hitachivantara.process;
+
+import com.google.common.base.Preconditions;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Semaphore;
+
+import static java.lang.String.format;
+
+/**
+ * A ProcessGovernor implementation which launches separate threads for each proc execution request,
+ * used for managing proc execution and completion.
+ *
+ *
+ * Current limitations
+ * 1)  Use any sort of defined queueing.  It's arbitrary which thread will be able to launch its proc when
+ * a permit becomes available.
+ * 2)  Support running procs under a different account.  Everything run under the permissions granted to the user
+ * running the JVM proc
+ * 3)  Support pooling/reuse of processes.
+ * 4)  Support any sort of process or session mgmt.  E.g. there's no way to kill or check status of procs invoked
+ *     from the governor.
+ * 5)  Limit threads.  Each .start() request will launch a thread use to acquire a permit, start the proc and
+ *     wait for results.  Consider replacing this design with an Actor pattern.
+ *
+ *
+ * Alternatives considered:
+ *  apache-commons-exec
+ *  https://github.com/zeroturnaround/zt-exec
+ *
+ *  Both have better support for IO handling, but don't pool or limit procs.
+ *
+ *  https://github.com/ViktorC/PP4J
+ *
+ *  Supports pooling, but seems primarily geared at external *java* processes.  Personal project of some guy.
+ */
+public class ProcessGovernorImpl implements ProcessGovernor {
+  private static final Logger logger = LoggerFactory.getLogger( ProcessGovernorImpl.class );
+  private final int maxProcesses;
+  private final ExecutorService execService;
+  private final Semaphore semaphore;
+
+  public ProcessGovernorImpl( ExecutorService execService, int maxProcesses ) {
```

Comment:
I'm wondering if the maxProcesses should just be baked in to the execService, and so we don't need to manage how many are running.  i.e. Executors.newFixedThreadPool().  So long as the thread waits for the process that we start to be completed, we would get all the pool logic for free.

Reply:
@kurtwalker - I think that would also mean blocking on the .start() call until a thread is available.  I had wanted to return a future right away so that the client can choose how to deal with waiting itself.

I'm thinking an actor pattern is appropriate here--i.e. have a queue that a single defined thread can manage--but didn't want to take the plunge on that until we had more motivation.

Reply:
good point.  would it help to use Future instead of CompleteableFuture?  then you could just directly return from the ExecuterService?

Reply:
I had been thinking it would be useful for the client to have a handle on the running Process, for IO, along with the ability to destroy the process directly.  If we just returned the Future<Process>, then the client wouldn't have a handle till the process is complete.